core/src/main/java/org/elasticsearch/index/shard/ShardPath.java

```diff
@@ -189,22 +190,41 @@ public static ShardPath selectNewPathForShard(NodeEnvironment env, ShardId shard
 
             // TODO - do we need something more extensible? Yet, this does the job for now...
             final NodeEnvironment.NodePath[] paths = env.nodePaths();
+
             NodeEnvironment.NodePath bestPath = null;
-            BigInteger maxUsableBytes = BigInteger.valueOf(Long.MIN_VALUE);
-            for (NodeEnvironment.NodePath nodePath : paths) {
-                FileStore fileStore = nodePath.fileStore;
 
-                BigInteger usableBytes = BigInteger.valueOf(fileStore.getUsableSpace());
-                assert usableBytes.compareTo(BigInteger.ZERO) >= 0;
+            if (paths.length == 1) {
+                // There's only one option, it's the best!
+                bestPath = paths[0];
+            } else {
+                int shardCount = indexSettings.getNumberOfShards();
+                // Maximum number of shards that a path should have for a particular index assuming
+                // all the shards were assigned to this node. For example, with a node with 4 data
+                // paths and an index with 9 primary shards, the maximum number of shards per path
+                // would be 3.
+                int maxShardsPerPath = Math.floorDiv(shardCount, paths.length) + ((shardCount % paths.length) == 0 ? 0 : 1);
+                ObjectLongHashMap<NodeEnvironment.NodePath> pathToShardCount = env.shardCountPerPath(shardId.getIndex());
 
-                // Deduct estimated reserved bytes from usable space:
-                Integer count = dataPathToShardCount.get(nodePath.path);
-                if (count != null) {
-                    usableBytes = usableBytes.subtract(estShardSizeInBytes.multiply(BigInteger.valueOf(count)));
-                }
-                if (bestPath == null || usableBytes.compareTo(maxUsableBytes) > 0) {
-                    maxUsableBytes = usableBytes;
-                    bestPath = nodePath;
+                BigInteger maxUsableBytes = BigInteger.valueOf(Long.MIN_VALUE);
+                for (NodeEnvironment.NodePath nodePath : paths) {
+                    FileStore fileStore = nodePath.fileStore;
+
+                    BigInteger usableBytes = BigInteger.valueOf(fileStore.getUsableSpace());
+                    assert usableBytes.compareTo(BigInteger.ZERO) >= 0;
+
+                    // Deduct estimated reserved bytes from usable space:
+                    Integer count = dataPathToShardCount.get(nodePath.path);
+                    if (count != null) {
+                        usableBytes = usableBytes.subtract(estShardSizeInBytes.multiply(BigInteger.valueOf(count)));
```

Comment:
I don't understand this. if we already have `count` shards allocated it should be already taken into account in usable space? why would we multiply it here again?

Reply:
This was a weighting that Mike originally added to make repeatedly assigned a shard to a given path weigh the pass less. I didn't change it (just used it verbatim).

I think it should be removed, however, in the subsequent work for this PR, since the space method will be rewritten from that perspective for that.